### PR TITLE
oneOf + allOf children - fix for union graph types and link objects

### PR DESCRIPTION
**Pull request by mzronek**
Dear maintainers,

I encountered a problem with the following OpenAPI pattern:

```
paths:
  '/composite':
    get:
      responses:
        '200':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Composite'
          links:
            testLink:
              operationId: anotherQuery
              parameters:
                someId: $request.path.id

```

```
components:
  schemas:
    Composite:
      oneOf:
        - $ref: '#/components/schemas/One'
        - $ref: '#/components/schemas/Two'
    Abstract:
      type: object
      properties:
        commonprop:
          type: object
    OneProps:
      type: object
      properties:
        differentoneprop:
          type: object
    TwoProps:
      type: object
      properties:
        differentoneprop:
          type: object
    One:
      allOf:
        - $ref: '#/components/schemas/Abstract'
        - $ref: '#/components/schemas/OneProps'
    Two:
      allOf:
        - $ref: '#/components/schemas/Abstract'
        - $ref: '#/components/schemas/TwoProps
```

There are currently two problems in the current implementation when using oneOf in combination with children using allOf:
* The target type can not be recognized as UNION and will fall back to JSON
* Any link objects will not work

This following PR fixes this for the API specifications I have in use. I had to move resolveAllOf to the oas tools file as I needed to resolve it when invoking GetOneOfTargetGraphQLType on each child node.

Please let me know if you have any questions.
Thanks for your hard work!

### CODE REVIEW

**Comment by mzronek** [comment]
Something failed in the Link checker stage.

```Successfully installed PyGithub-1.53 certifi-2020.6.20 chardet-3.0.4 deprecated-1.2.10 idna-2.10 pyjwt-1.7.1 requests-2.24.0 setuptools-49.6.0 urllib3-1.25.10 wheel-0.35.1 wrapt-1.12.1
/opt/hostedtoolcache/Python/3.9.7/x64/bin/python /home/runner/work/_actions/peter-evans/create-issue-from-file/v2/dist/ciff/create_issue_from_file.py
Traceback (most recent call last):
  File "/home/runner/work/_actions/peter-evans/create-issue-from-file/v2/dist/ciff/create_issue_from_file.py", line 72, in <module>
    issue = repo.create_issue(title=issue_title, body=issue_content)
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/site-packages/github/Repository.py", line 1213, in create_issue
    headers, data = self._requester.requestJsonAndCheck(
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/site-packages/github/Requester.py", line 317, in requestJsonAndCheck
    return self.__check(
  File "/opt/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/site-packages/github/Requester.py", line 342, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.GithubException: 403 {"message": "Resource not accessible by integration", "documentation_url": "https://docs.github.com/rest/reference/issues#create-an-issue"}
Error: The process '/opt/hostedtoolcache/Python/3.9.7/x64/bin/python' failed with exit code 1```

Is there something missing from my side or anything I can do to help?

**Comment by Alan-Cha** [comment]
@mzronek Sorry for the very long wait... I did take a look initially but I was confused by a few things and then I got caught up in other work and now here we are... 😓 

But, I've taken another look and I've gotten to fully appreciate everything you've done. This is a super cool change!!! I didn't think that oneOfs could have link objects that can be trickled down to its member types.

File: packages/openapi-to-graphql/test/fixtures/example_oas4.json [hunk 1]
```diff
@@ -533,6 +574,60 @@
...
+        "type": "object",
+        "properties": {
+          "commonProp": {
+            "type": "string"
+          }
+        }
+      },
+      "OneProps": {
+        "type": "object",
+        "properties": {
```
Comment (Alan-Cha) [inline packages/openapi-to-graphql/test/fixtures/example_oas4.json:598]: I was initially confused as to what the expected schema is from your OAS. I was expecting the member types to be object types but they were JSON instead. I thought there may have been an issue with the code but I later realized that your example was more for illustration. By giving the member types `properties`, they no longer resolve to JSON and your example become clearer.

File: packages/openapi-to-graphql/test/fixtures/example_oas4.json [hunk 2]
```diff
@@ -506,6 +506,47 @@
...
+            "description": "Return a user.\n\nEquivalent to GET /oneOfWithAllOfsAndLink",
+            "content": {
+              "application/json": {
+                "schema": {
+                  "$ref": "#/components/schemas/OneOfWithAllOfsAndLink"
+                }
+              }
+            },
+            "links": {
+              "testLink": {
```
Comment (Alan-Cha) [inline packages/openapi-to-graphql/test/fixtures/example_oas4.json:540]: I was also confused why the link field wasn't being produced.

File: packages/openapi-to-graphql/src/preprocessor.ts [hunk 3]
```diff
@@ -699,11 +699,29 @@ export function createDataDef<TSource, TContext, TArgs>(
...
-      additionalLinks: saneLinks,
-      existingDataDef,
-      data
-    })
+    /**
+     * Special handling for oneOf. Subdefinitions are always an array 
+     * (see createOneOfUnion)
+     */
+    if (
+      existingDataDef.targetGraphQLType === TargetGraphQLType.oneOfUnion &&
```
Comment (Alan-Cha) [inline packages/openapi-to-graphql/src/preprocessor.ts:707]: I see what you were trying to do here but actually the OAS you provided doesn't make use of the code here. This is to collapse links in existing data definitions.

File: packages/openapi-to-graphql/src/preprocessor.ts [hunk 4]
```diff
@@ -1672,7 +1562,8 @@ function createOneOfUnion<TSource, TContext, TArgs>(
       memberSchema,
       isInputObjectType,
       data,
-      oas
+      oas,
+      def.links
```
Comment (Alan-Cha) [inline packages/openapi-to-graphql/src/preprocessor.ts:1566]: This line `def.links` allows OtG to trickle links from oneOf to its member schemas during initial creation of the data definitions, as opposed to collapsing links in existing data definitions from the previous comment.

**Comment by Alan-Cha** [comment]
Anyway, thank you so much for these changes!!! I am very happy that we finally got these in 😄

**Pull request merged**